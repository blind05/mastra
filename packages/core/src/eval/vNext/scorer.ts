export interface ScorerResult {
  id: string;
  name: string;
  traceId: string;
  runId: string;
  evaluator: string;
  result: Record<string, any>;
  metadata: Record<string, any>;
  input: string;
  output: string;
  additionalContext: string;
  runtimeContext: Record<string, any>;
  /**
   * Things you can evaluate
   */
  entityType: 'AGENT' | 'WORKFLOW';
  entity: Record<string, any>;
  source: string;
  resourceId: string;
  threadId: string;
  createdAt: Date;
  updatedAt: Date;
}

export type ScorerInput = {
  input: string;
  output: string;
  entityId: string;
  entityType: 'AGENT' | 'WORKFLOW';
  entity: Record<string, any>;
  runId: string;
};

export abstract class Scorer {
  abstract evaluate({ input, output, entityId, entityType, entity, runId }: ScorerInput): Promise<ScorerResult>;
}
