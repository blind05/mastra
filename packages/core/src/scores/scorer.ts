// ================================
// CORE TYPES
// ================================

import type { LanguageModel } from '../llm';
import { z } from 'zod';

// Pipeline metadata
interface ScorerConfig {
  name: string;
  description: string;
  judge?: {
    model: LanguageModel;
    instructions: string;
  };
}

// Standardized input type for all pipelines
interface ScorerRun {
  runId?: string;
  input?: Record<string, any>[];
  output: Record<string, any>;
  runtimeContext?: Record<string, any>;
}

// Helper types
type StepResultKey<T extends string> = `${T}StepResult`;
type AccumulatedResults<T extends Record<string, any>, K extends string, V> = T & Record<StepResultKey<K>, V>;
type StepContext<TAccumulated extends Record<string, any>, TRun> = {
  run: TRun;
  results: TAccumulated;
};

// Special context type for generateReason that includes the score
type GenerateReasonContext<TAccumulated extends Record<string, any>> = StepContext<TAccumulated, ScorerRun> & {
  score: TAccumulated extends Record<'generateScoreStepResult', infer TScore> ? TScore : never;
};

// Conditional type for PromptObject context
type PromptObjectContext<
  TAccumulated extends Record<string, any>,
  TStepName extends string,
> = TStepName extends 'generateReason' ? GenerateReasonContext<TAccumulated> : StepContext<TAccumulated, ScorerRun>;

// Prompt object definition with conditional typing
interface PromptObject<TOutput, TAccumulated extends Record<string, any>, TStepName extends string = string> {
  description: string;
  outputSchema: z.ZodSchema<TOutput>;
  createPrompt: (context: PromptObjectContext<TAccumulated, TStepName>) => string;
}

class MastraNewScorer<TAccumulatedResults extends Record<string, any> = {}> {
  constructor(
    private metadata: ScorerConfig,
    private steps: Array<{
      name: string;
      execute: (context: any) => any;
      isPromptObject: boolean;
      description?: string;
    }> = [],
    private originalPromptObjects: Map<string, PromptObject<any, any, any>> = new Map(),
  ) {}

  // Getter for pipeline name
  get name(): string {
    return this.metadata.name;
  }

  // Getter for pipeline description
  get description(): string {
    return this.metadata.description;
  }

  // Check if generateScore exists
  private get hasGenerateScore(): boolean {
    return this.steps.some(step => step.name === 'generateScore');
  }

  preprocess<TPreprocessOutput>(
    stepDef:
      | ((context: StepContext<TAccumulatedResults, ScorerRun>) => TPreprocessOutput)
      | PromptObject<TPreprocessOutput, TAccumulatedResults, 'preprocess'>,
  ): MastraNewScorer<AccumulatedResults<TAccumulatedResults, 'preprocess', TPreprocessOutput>> {
    const isPromptObj = this.isPromptObject(stepDef);

    if (isPromptObj) {
      const promptObj = stepDef as PromptObject<TPreprocessOutput, TAccumulatedResults, 'preprocess'>;
      this.originalPromptObjects.set('preprocess', promptObj);
    }

    return new MastraNewScorer(
      this.metadata,
      [
        ...this.steps,
        {
          name: 'preprocess',
          execute: isPromptObj
            ? this.createPromptExecutor(stepDef as PromptObject<TPreprocessOutput, TAccumulatedResults, 'preprocess'>)
            : (stepDef as (context: StepContext<any, ScorerRun>) => TPreprocessOutput),
          isPromptObject: isPromptObj,
          description: isPromptObj
            ? (stepDef as PromptObject<TPreprocessOutput, TAccumulatedResults, 'preprocess'>).description
            : undefined,
        },
      ],
      new Map(this.originalPromptObjects),
    );
  }

  analyze<TAnalyzeOutput>(
    stepDef:
      | ((context: StepContext<TAccumulatedResults, ScorerRun>) => TAnalyzeOutput)
      | PromptObject<TAnalyzeOutput, TAccumulatedResults, 'analyze'>,
  ): MastraNewScorer<AccumulatedResults<TAccumulatedResults, 'analyze', TAnalyzeOutput>> {
    const isPromptObj = this.isPromptObject(stepDef);

    if (isPromptObj) {
      const promptObj = stepDef as PromptObject<TAnalyzeOutput, TAccumulatedResults, 'analyze'>;
      this.originalPromptObjects.set('analyze', promptObj);
    }

    return new MastraNewScorer(
      this.metadata,
      [
        ...this.steps,
        {
          name: 'analyze',
          execute: isPromptObj
            ? this.createPromptExecutor(stepDef as PromptObject<TAnalyzeOutput, TAccumulatedResults, 'analyze'>)
            : (stepDef as (context: StepContext<any, ScorerRun>) => TAnalyzeOutput),
          isPromptObject: isPromptObj,
          description: isPromptObj
            ? (stepDef as PromptObject<TAnalyzeOutput, TAccumulatedResults, 'analyze'>).description
            : undefined,
        },
      ],
      new Map(this.originalPromptObjects),
    );
  }

  generateScore<TScoreOutput extends number>(
    stepDef:
      | ((context: StepContext<TAccumulatedResults, ScorerRun>) => TScoreOutput)
      | PromptObject<TScoreOutput, TAccumulatedResults, 'generateScore'>,
  ): MastraNewScorer<AccumulatedResults<TAccumulatedResults, 'generateScore', TScoreOutput>> {
    const isPromptObj = this.isPromptObject(stepDef);

    if (isPromptObj) {
      const promptObj = stepDef as PromptObject<TScoreOutput, TAccumulatedResults, 'generateScore'>;
      this.originalPromptObjects.set('generateScore', promptObj);
    }

    return new MastraNewScorer(
      this.metadata,
      [
        ...this.steps,
        {
          name: 'generateScore',
          execute: isPromptObj
            ? this.createPromptExecutor(stepDef as PromptObject<TScoreOutput, TAccumulatedResults, 'generateScore'>)
            : (stepDef as (context: StepContext<any, ScorerRun>) => TScoreOutput),
          isPromptObject: isPromptObj,
          description: isPromptObj
            ? (stepDef as PromptObject<TScoreOutput, TAccumulatedResults, 'generateScore'>).description
            : undefined,
        },
      ],
      new Map(this.originalPromptObjects),
    );
  }

  generateReason<TReasonOutput>(
    stepDef:
      | ((context: GenerateReasonContext<TAccumulatedResults>) => TReasonOutput)
      | PromptObject<TReasonOutput, TAccumulatedResults, 'generateReason'>,
  ): MastraNewScorer<AccumulatedResults<TAccumulatedResults, 'generateReason', TReasonOutput>> {
    // Runtime check: generateReason only allowed after generateScore
    if (!this.hasGenerateScore) {
      throw new Error(`Pipeline "${this.metadata.name}": generateReason() can only be called after generateScore()`);
    }

    const isPromptObj = this.isPromptObject(stepDef);

    if (isPromptObj) {
      const promptObj = stepDef as PromptObject<TReasonOutput, TAccumulatedResults, 'generateReason'>;
      this.originalPromptObjects.set('generateReason', promptObj);
    }

    return new MastraNewScorer(
      this.metadata,
      [
        ...this.steps,
        {
          name: 'generateReason',
          execute: isPromptObj
            ? this.createPromptExecutor(stepDef as PromptObject<TReasonOutput, TAccumulatedResults, 'generateReason'>)
            : (stepDef as (context: GenerateReasonContext<any>) => TReasonOutput),
          isPromptObject: isPromptObj,
          description: isPromptObj
            ? (stepDef as PromptObject<TReasonOutput, TAccumulatedResults, 'generateReason'>).description
            : undefined,
        },
      ],
      new Map(this.originalPromptObjects),
    );
  }

  run(
    input: ScorerRun,
    options: {
      llmCall?: (prompt: string, schema: z.ZodSchema<any>) => any;
      logPrompts?: boolean;
    } = {},
  ): {
    input: ScorerRun;
    results: TAccumulatedResults;
    finalResult: any;
    generatedPrompts: Array<{ stepName: string; prompt: string; description: string }>;
    score: TAccumulatedResults extends Record<'generateScoreStepResult', infer TScore> ? TScore : never;
    reason?: TAccumulatedResults extends Record<'generateReasonStepResult', infer TReason> ? TReason : undefined;
  } {
    // Runtime check: execute only allowed after generateScore
    if (!this.hasGenerateScore) {
      throw new Error(
        `Pipeline "${this.metadata.name}": Cannot execute pipeline without generateScore() step. ` +
          `Current steps: [${this.steps.map(s => s.name).join(', ')}]`,
      );
    }

    let accumulatedResults: Record<string, any> = {};
    let lastStepResult: any = null;
    const generatedPrompts: Array<{ stepName: string; prompt: string; description: string }> = [];

    console.log(`üöÄ Starting pipeline "${this.metadata.name}" [${input.runId || 'no-id'}]`);

    for (const step of this.steps) {
      const startTime = performance.now();

      try {
        // Create context based on step type
        const context =
          step.name === 'generateReason'
            ? {
                run: input,
                results: accumulatedResults,
                score: accumulatedResults.generateScoreStepResult,
              }
            : {
                run: input,
                results: accumulatedResults,
              };

        let stepResult: any;

        if (step.isPromptObject) {
          const originalStep = this.originalPromptObjects.get(step.name);
          if (originalStep) {
            const prompt = originalStep.createPrompt(context);

            generatedPrompts.push({
              stepName: step.name,
              prompt: prompt,
              description: originalStep.description,
            });

            if (options.logPrompts) {
              console.log(`üìù Prompt for "${step.name}":\n${prompt}`);
            }

            if (options.llmCall) {
              stepResult = options.llmCall(prompt, originalStep.outputSchema);
            } else {
              stepResult = this.generateMockResponse(originalStep.outputSchema, step.name);
            }

            stepResult = originalStep.outputSchema.parse(stepResult);
          }
        } else {
          stepResult = step.execute(context);
        }

        const duration = performance.now() - startTime;
        const resultKey = `${step.name}StepResult`;
        accumulatedResults[resultKey] = stepResult;
        lastStepResult = stepResult;

        console.log(`‚úÖ ${step.name} (${duration.toFixed(1)}ms)`);
      } catch (error) {
        console.error(`‚ùå Step "${step.name}" failed:`, error);
        throw new Error(`Pipeline execution failed at step: ${step.name}`);
      }
    }

    console.log(`üèÅ Pipeline "${this.metadata.name}" completed`);

    return {
      input: input,
      results: accumulatedResults as TAccumulatedResults,
      finalResult: lastStepResult,
      generatedPrompts: generatedPrompts,
      score: accumulatedResults.generateScoreStepResult,
      reason: accumulatedResults.generateReasonStepResult,
    };
  }

  private isPromptObject(stepDef: any): boolean {
    return (
      typeof stepDef === 'object' && 'description' in stepDef && 'outputSchema' in stepDef && 'createPrompt' in stepDef
    );
  }

  private createPromptExecutor<T>(promptObj: PromptObject<T, any, any>) {
    return (context: any): T => {
      return this.generateMockResponse(promptObj.outputSchema, 'mock');
    };
  }

  private generateMockResponse(schema: z.ZodSchema<any>, stepName: string): any {
    if (schema instanceof z.ZodObject) {
      const shape = schema._def.shape();
      const mockObj: any = {};

      for (const [key, fieldSchema] of Object.entries(shape)) {
        const field = fieldSchema as z.ZodTypeAny;

        if (field instanceof z.ZodString) {
          mockObj[key] = `Mock ${key}`;
        } else if (field instanceof z.ZodNumber) {
          mockObj[key] = Math.floor(Math.random() * 100);
        } else if (field instanceof z.ZodBoolean) {
          mockObj[key] = Math.random() > 0.5;
        } else if (field instanceof z.ZodArray) {
          mockObj[key] = [`Item 1`, `Item 2`];
        } else if (field instanceof z.ZodEnum) {
          const values = (field as any)._def.values;
          mockObj[key] = values[Math.floor(Math.random() * values.length)];
        } else {
          mockObj[key] = `Mock ${key}`;
        }
      }

      return mockObj;
    }

    return `Mock response for ${stepName}`;
  }

  getSteps(): Array<{ name: string; type: 'function' | 'prompt'; description?: string }> {
    return this.steps.map(step => ({
      name: step.name,
      type: step.isPromptObject ? 'prompt' : 'function',
      description: step.description,
    }));
  }

  // Get complete pipeline metadata including steps
  getMetadata(): {
    name: string;
    description: string;
    steps: Array<{ name: string; type: 'function' | 'prompt'; description?: string }>;
    stepCount: number;
    hasGenerateScore: boolean;
  } {
    return {
      name: this.metadata.name,
      description: this.metadata.description,
      steps: this.getSteps(),
      stepCount: this.steps.length,
      hasGenerateScore: this.hasGenerateScore,
    };
  }
}

export function createNewScorer({ name, description, judge }: ScorerConfig): MastraNewScorer<{}> {
  return new MastraNewScorer<{}>({ name, description, judge });
}
