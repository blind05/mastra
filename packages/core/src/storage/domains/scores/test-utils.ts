import { randomUUID } from 'crypto';
import { describe, it, expect } from 'vitest';
import type { ScoreRowData } from '../../../eval';
import type { MastraStorage } from '../../base';

export function createMockScore({ scorerId }: { scorerId: string }): ScoreRowData {
  return {
    id: randomUUID(),
    entityId: 'eval-agent',
    entityType: 'AGENT',
    scorerId,
    createdAt: new Date(),
    updatedAt: new Date(),
    runId: 'run-123',
    reason: 'Sample reason',
    extractStepResult: {
      text: 'Sample extract step result',
    },
    analyzeStepResult: {
      text: 'Sample analyze step result',
    },
    score: 0.8,
    extractPrompt: 'Sample extract prompt',
    analyzePrompt: 'Sample analyze prompt',
    reasonPrompt: 'Sample reason prompt',
    scorer: {
      id: scorerId,
      name: 'my-eval',
      description: 'My eval',
    },
    input: [
      {
        id: randomUUID(),
        name: 'input-1',
        value: 'Sample input',
      },
    ],
    output: {
      text: 'Sample output',
    },
    source: 'LIVE',
    entity: {
      id: 'eval-agent',
      name: 'Sample entity',
    },
    runtimeContext: {},
  };
}

export function createScoresTest({ storage }: { storage: MastraStorage }) {
  describe('Score Operations', () => {
    it('should retrieve scores by scorer id', async () => {
      const scorerId = `scorer-${randomUUID()}`;

      // Create sample scores
      const score1 = createMockScore({ scorerId });
      const score2 = createMockScore({ scorerId });
      const score3 = createMockScore({ scorerId });

      // Insert evals

      await storage.saveScore(score1);
      await storage.saveScore(score2);
      await storage.saveScore(score3);

      // Test getting all evals for the agent
      const allScoresByScorerId = await storage.getScoresByScorerId({ scorerId, pagination: { page: 0, perPage: 10 } });
      expect(allScoresByScorerId?.scores).toHaveLength(3);
      expect(allScoresByScorerId?.scores.map(e => e.runId)).toEqual(
        expect.arrayContaining([score1.runId, score2.runId, score3.runId]),
      );

      // Test getting scores for non-existent scorer
      const nonExistentScores = await storage.getScoresByScorerId({
        scorerId: 'non-existent-scorer',
        pagination: { page: 0, perPage: 10 },
      });
      expect(nonExistentScores?.scores).toHaveLength(0);
    });
  });
}
