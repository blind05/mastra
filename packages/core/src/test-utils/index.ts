export { createMockModel, MockProvider } from './llm-mock';
export { MockLanguageModelV2 } from 'ai/test';
export { simulateReadableStream } from 'ai';
