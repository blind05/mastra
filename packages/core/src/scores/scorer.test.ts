import { FunctionBasedScorerBuilders } from './scorer.test-utils';
import { describe, it, expect, beforeEach, vi } from 'vitest';
import { MockLanguageModelV1 } from 'ai/test';
import { createNewScorer } from './scorer';
import z from 'zod';

export const PromptBasedScorerBuilders = {
  withAnalyze: createNewScorer({
    name: 'test-scorer',
    description: 'A test scorer',
  })
    .analyze({
      judge: {
        model: new MockLanguageModelV1({
          defaultObjectGenerationMode: 'json',
          doGenerate: async () => ({
            rawCall: { rawPrompt: null, rawSettings: {} },
            finishReason: 'stop',
            usage: { promptTokens: 10, completionTokens: 20 },
            text: `{
                      "inputLength": 10,
                      "outputLength": 11
                  }`,
          }),
        }),
        instructions: `Test instructions`,
      },
      description: 'Analyze the input and output',
      outputSchema: z.object({
        inputLength: z.number(),
        outputLength: z.number(),
      }),
      createPrompt: () => {
        return `Test Analyze prompt`;
      },
    })
    .generateScore(({ results }) => {
      const inputLength = results.analyzeStepResult?.inputLength;
      const outputLength = results.analyzeStepResult?.outputLength;
      return inputLength !== undefined && outputLength !== undefined ? 1 : 0;
    }),

  withPreprocessAndAnalyze: createNewScorer({
    name: 'test-scorer',
    description: 'A test scorer',
  })
    .preprocess({
      judge: {
        model: new MockLanguageModelV1({
          defaultObjectGenerationMode: 'json',
          doGenerate: async () => ({
            rawCall: { rawPrompt: null, rawSettings: {} },
            finishReason: 'stop',
            usage: { promptTokens: 10, completionTokens: 20 },
            text: `{
                            "reformattedInput": "TEST INPUT",
                            "reformattedOutput": "TEST OUTPUT"
                        }`,
          }),
        }),
        instructions: `Test instructions`,
      },
      description: 'Preprocess the input and output',
      outputSchema: z.object({
        reformattedInput: z.string(),
        reformattedOutput: z.string(),
      }),
      createPrompt: () => {
        return `Test Preprocess prompt`;
      },
    })
    .analyze({
      description: 'Analyze the input and output',
      outputSchema: z.object({
        inputLength: z.number(),
        outputLength: z.number(),
      }),
      createPrompt: () => {
        return `Test Analyze prompt`;
      },
      judge: {
        model: new MockLanguageModelV1({
          defaultObjectGenerationMode: 'json',
          doGenerate: async () => ({
            rawCall: { rawPrompt: null, rawSettings: {} },
            finishReason: 'stop',
            usage: { promptTokens: 10, completionTokens: 20 },
            text: `{
                        "inputLength": 10,
                        "outputLength": 11
                    }`,
          }),
        }),
        instructions: `Test instructions`,
      },
    })
    .generateScore(({ results }) => {
      const inputLength = results.analyzeStepResult?.inputLength;
      const outputLength = results.analyzeStepResult?.outputLength;
      return inputLength !== undefined && outputLength !== undefined ? 1 : 0;
    }),

  withAnalyzeAndReason: createNewScorer({
    name: 'test-scorer',
    description: 'A test scorer',
  })
    .analyze({
      description: 'Analyze the input and output',
      outputSchema: z.object({
        inputLength: z.number(),
        outputLength: z.number(),
      }),
      createPrompt: () => {
        return `Test Analyze prompt`;
      },
      judge: {
        model: new MockLanguageModelV1({
          defaultObjectGenerationMode: 'json',
          doGenerate: async () => ({
            rawCall: { rawPrompt: null, rawSettings: {} },
            finishReason: 'stop',
            usage: { promptTokens: 10, completionTokens: 20 },
            text: `{
                        "inputLength": 10,
                        "outputLength": 11
                    }`,
          }),
        }),
        instructions: `Test instructions`,
      },
    })
    .generateScore(({ results }) => {
      const inputLength = results.analyzeStepResult?.inputLength;
      const outputLength = results.analyzeStepResult?.outputLength;
      return inputLength !== undefined && outputLength !== undefined ? 1 : 0;
    })
    .generateReason({
      judge: {
        model: new MockLanguageModelV1({
          defaultObjectGenerationMode: 'json',
          doGenerate: async () => ({
            rawCall: { rawPrompt: null, rawSettings: {} },
            finishReason: 'stop',
            usage: { promptTokens: 10, completionTokens: 20 },
            text: `This is a test reason`,
          }),
        }),
        instructions: `Test instructions`,
      },
      description: 'Generate a reason for the score',
      createPrompt: () => {
        return `Test Generate Reason prompt`;
      },
    }),

  withAllSteps: createNewScorer({
    name: 'test-scorer',
    description: 'A test scorer',
  })
    .preprocess({
      judge: {
        model: new MockLanguageModelV1({
          defaultObjectGenerationMode: 'json',
          doGenerate: async () => ({
            rawCall: { rawPrompt: null, rawSettings: {} },
            finishReason: 'stop',
            usage: { promptTokens: 10, completionTokens: 20 },
            text: `{
                            "reformattedInput": "TEST INPUT",
                            "reformattedOutput": "TEST OUTPUT"
                        }`,
          }),
        }),
        instructions: `Test instructions`,
      },
      description: 'Preprocess the input and output',
      outputSchema: z.object({
        reformattedInput: z.string(),
        reformattedOutput: z.string(),
      }),
      createPrompt: () => {
        return `Test Preprocess prompt`;
      },
    })
    .analyze({
      description: 'Analyze the input and output',
      outputSchema: z.object({
        inputLength: z.number(),
        outputLength: z.number(),
      }),
      createPrompt: () => {
        return `Test Analyze prompt`;
      },
      judge: {
        model: new MockLanguageModelV1({
          defaultObjectGenerationMode: 'json',
          doGenerate: async () => ({
            rawCall: { rawPrompt: null, rawSettings: {} },
            finishReason: 'stop',
            usage: { promptTokens: 10, completionTokens: 20 },
            text: `{
                        "inputLength": 10,
                        "outputLength": 11
                    }`,
          }),
        }),
        instructions: `Test instructions`,
      },
    })
    .generateScore(({ results }) => {
      const inputLength = results.analyzeStepResult?.inputLength;
      const outputLength = results.analyzeStepResult?.outputLength;
      return inputLength !== undefined && outputLength !== undefined ? 1 : 0;
    })
    .generateReason({
      judge: {
        model: new MockLanguageModelV1({
          defaultObjectGenerationMode: 'json',
          doGenerate: async () => ({
            rawCall: { rawPrompt: null, rawSettings: {} },
            finishReason: 'stop',
            usage: { promptTokens: 10, completionTokens: 20 },
            text: `{
                            "reason": "TEST REASON"
                        }`,
          }),
        }),
        instructions: `Test instructions`,
      },
      description: 'Generate a reason for the score',
      createPrompt: () => {
        return `Test Generate Reason prompt`;
      },
    }),
};

const createTestData = () => ({
  inputText: 'test input',
  outputText: 'test output',
  get userInput() {
    return [{ role: 'user', content: this.inputText }];
  },
  get agentOutput() {
    return { role: 'assistant', text: this.outputText };
  },
  get scoringInput() {
    return { input: this.userInput, output: this.agentOutput };
  },
});

describe('createScorer', () => {
  let testData: ReturnType<typeof createTestData>;

  beforeEach(() => {
    testData = createTestData();
  });

  describe('Steps as functions scorer', () => {
    it('should create a basic scorer with functions', async () => {
      const scorer = FunctionBasedScorerBuilders.basic;
      const result = await scorer.run(testData.scoringInput);

      expect(result).toMatchSnapshot();
    });

    it('should create a scorer with reason', async () => {
      const scorer = FunctionBasedScorerBuilders.withReason;
      const result = await scorer.run(testData.scoringInput);

      expect(result).toMatchSnapshot();
    });

    it('should create a scorer with preprocess and reason', async () => {
      const scorer = FunctionBasedScorerBuilders.withPreprocessAndReason;
      const result = await scorer.run(testData.scoringInput);

      expect(result).toMatchSnapshot();
    });

    it('should create a scorer with preprocess and analyze', async () => {
      const scorer = FunctionBasedScorerBuilders.withPreprocessAndAnalyze;
      const result = await scorer.run(testData.scoringInput);

      expect(result).toMatchSnapshot();
    });

    it('should create a scorer with preprocess only', async () => {
      const scorer = FunctionBasedScorerBuilders.withPreprocess;
      const result = await scorer.run(testData.scoringInput);

      expect(result).toMatchSnapshot();
    });

    it('should create a scorer with preprocess, analyze, and reason', async () => {
      const scorer = FunctionBasedScorerBuilders.withPreprocessAndAnalyzeAndReason;
      const result = await scorer.run(testData.scoringInput);

      expect(result).toMatchSnapshot();
    });

    it('should create a scorer with analyze only', async () => {
      const scorer = FunctionBasedScorerBuilders.withAnalyze;
      const result = await scorer.run(testData.scoringInput);

      expect(result).toMatchSnapshot();
    });

    it('should create a scorer with analyze and reason', async () => {
      const scorer = FunctionBasedScorerBuilders.withAnalyzeAndReason;
      const result = await scorer.run(testData.scoringInput);

      expect(result).toMatchSnapshot();
    });
  });

  describe('Steps as prompt objects scorer', () => {
    it('with analyze prompt object', async () => {
      const scorer = PromptBasedScorerBuilders.withAnalyze;
      const result = await scorer.run(testData.scoringInput);

      expect(result).toMatchSnapshot();
    });

    it('with preprocess and analyze prompt object', async () => {
      const scorer = PromptBasedScorerBuilders.withPreprocessAndAnalyze;

      const result = await scorer.run(testData.scoringInput);

      expect(result).toMatchSnapshot();
    });

    it('with analyze and reason prompt object', async () => {
      const scorer = PromptBasedScorerBuilders.withAnalyzeAndReason;
      const result = await scorer.run(testData.scoringInput);

      expect(typeof result.reason).toBe('string');
      expect(result).toMatchSnapshot();
    });
  });
});
