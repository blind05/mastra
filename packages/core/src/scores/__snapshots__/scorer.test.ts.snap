// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`createScorer > Steps as functions scorer > should create a basic scorer with functions 1`] = `
{
  "analyzePrompt": undefined,
  "analyzeResult": undefined,
  "generateReasonPrompt": undefined,
  "generateScorePrompt": undefined,
  "preprocessPrompt": undefined,
  "preprocessResult": undefined,
  "reason": undefined,
  "run": {
    "input": [
      {
        "content": "test input",
        "role": "user",
      },
    ],
    "output": {
      "role": "assistant",
      "text": "test output",
    },
  },
  "score": 1,
}
`;

exports[`createScorer > Steps as functions scorer > should create a scorer with analyze and reason 1`] = `
{
  "analyzePrompt": undefined,
  "analyzeResult": {
    "inputFromAnalyze": "test input!",
    "outputFromAnalyze": "test output!",
  },
  "generateReasonPrompt": undefined,
  "generateScorePrompt": undefined,
  "preprocessPrompt": undefined,
  "preprocessResult": undefined,
  "reason": "the reason the score is 1 is because the input is test input! and the output is test output!",
  "run": {
    "input": [
      {
        "content": "test input",
        "role": "user",
      },
    ],
    "output": {
      "role": "assistant",
      "text": "test output",
    },
  },
  "score": 1,
}
`;

exports[`createScorer > Steps as functions scorer > should create a scorer with analyze only 1`] = `
{
  "analyzePrompt": undefined,
  "analyzeResult": {
    "inputFromAnalyze": "test input!",
    "outputFromAnalyze": "test output!",
  },
  "generateReasonPrompt": undefined,
  "generateScorePrompt": undefined,
  "preprocessPrompt": undefined,
  "preprocessResult": undefined,
  "reason": undefined,
  "run": {
    "input": [
      {
        "content": "test input",
        "role": "user",
      },
    ],
    "output": {
      "role": "assistant",
      "text": "test output",
    },
  },
  "score": 1,
}
`;

exports[`createScorer > Steps as functions scorer > should create a scorer with preprocess and analyze 1`] = `
{
  "analyzePrompt": undefined,
  "analyzeResult": {
    "inputFromAnalyze": "TEST INPUT!",
    "outputFromAnalyze": "TEST OUTPUT!",
  },
  "generateReasonPrompt": undefined,
  "generateScorePrompt": undefined,
  "preprocessPrompt": undefined,
  "preprocessResult": {
    "reformattedInput": "TEST INPUT",
    "reformattedOutput": "TEST OUTPUT",
  },
  "reason": undefined,
  "run": {
    "input": [
      {
        "content": "test input",
        "role": "user",
      },
    ],
    "output": {
      "role": "assistant",
      "text": "test output",
    },
  },
  "score": 1,
}
`;

exports[`createScorer > Steps as functions scorer > should create a scorer with preprocess and reason 1`] = `
{
  "analyzePrompt": undefined,
  "analyzeResult": undefined,
  "generateReasonPrompt": undefined,
  "generateScorePrompt": undefined,
  "preprocessPrompt": undefined,
  "preprocessResult": {
    "reformattedInput": "TEST INPUT",
    "reformattedOutput": "TEST OUTPUT",
  },
  "reason": "the reason the score is 1 is because the input is TEST INPUT and the output is TEST OUTPUT",
  "run": {
    "input": [
      {
        "content": "test input",
        "role": "user",
      },
    ],
    "output": {
      "role": "assistant",
      "text": "test output",
    },
  },
  "score": 1,
}
`;

exports[`createScorer > Steps as functions scorer > should create a scorer with preprocess only 1`] = `
{
  "analyzePrompt": undefined,
  "analyzeResult": undefined,
  "generateReasonPrompt": undefined,
  "generateScorePrompt": undefined,
  "preprocessPrompt": undefined,
  "preprocessResult": {
    "reformattedInput": "TEST INPUT",
    "reformattedOutput": "TEST OUTPUT",
  },
  "reason": undefined,
  "run": {
    "input": [
      {
        "content": "test input",
        "role": "user",
      },
    ],
    "output": {
      "role": "assistant",
      "text": "test output",
    },
  },
  "score": 1,
}
`;

exports[`createScorer > Steps as functions scorer > should create a scorer with preprocess, analyze, and reason 1`] = `
{
  "analyzePrompt": undefined,
  "analyzeResult": {
    "inputFromAnalyze": "TEST INPUT!",
    "outputFromAnalyze": "TEST OUTPUT!",
  },
  "generateReasonPrompt": undefined,
  "generateScorePrompt": undefined,
  "preprocessPrompt": undefined,
  "preprocessResult": {
    "reformattedInput": "TEST INPUT",
    "reformattedOutput": "TEST OUTPUT",
  },
  "reason": "the reason the score is 1 is because the input is TEST INPUT! and the output is TEST OUTPUT!",
  "run": {
    "input": [
      {
        "content": "test input",
        "role": "user",
      },
    ],
    "output": {
      "role": "assistant",
      "text": "test output",
    },
  },
  "score": 1,
}
`;

exports[`createScorer > Steps as functions scorer > should create a scorer with reason 1`] = `
{
  "analyzePrompt": undefined,
  "analyzeResult": undefined,
  "generateReasonPrompt": undefined,
  "generateScorePrompt": undefined,
  "preprocessPrompt": undefined,
  "preprocessResult": undefined,
  "reason": "the reason the score is 1 is because the input is test input and the output is test output",
  "run": {
    "input": [
      {
        "content": "test input",
        "role": "user",
      },
    ],
    "output": {
      "role": "assistant",
      "text": "test output",
    },
  },
  "score": 1,
}
`;

exports[`createScorer > Steps as prompt objects scorer > should create a scorer with analyze prompt object 1`] = `
{
  "analyzePrompt": "Test Analyze prompt",
  "analyzeResult": {
    "inputLength": 27,
    "outputLength": 86,
  },
  "generateReasonPrompt": undefined,
  "generateScorePrompt": undefined,
  "preprocessPrompt": undefined,
  "preprocessResult": undefined,
  "reason": undefined,
  "run": {
    "input": [
      {
        "content": "test input",
        "role": "user",
      },
    ],
    "output": {
      "role": "assistant",
      "text": "test output",
    },
  },
  "score": 1,
}
`;

exports[`createScorer > Steps as prompt objects scorer > with analyze and reason prompt object 1`] = `
{
  "analyzePrompt": "Test Analyze prompt",
  "analyzeResult": {
    "inputLength": 10,
    "outputLength": 11,
  },
  "generateReasonPrompt": "Test Generate Reason prompt",
  "generateScorePrompt": undefined,
  "preprocessPrompt": undefined,
  "preprocessResult": undefined,
  "reason": "This is a test reason",
  "run": {
    "input": [
      {
        "content": "test input",
        "role": "user",
      },
    ],
    "output": {
      "role": "assistant",
      "text": "test output",
    },
  },
  "score": 1,
}
`;

exports[`createScorer > Steps as prompt objects scorer > with analyze prompt object 1`] = `
{
  "analyzePrompt": "Test Analyze prompt",
  "analyzeResult": {
    "inputLength": 10,
    "outputLength": 11,
  },
  "generateReasonPrompt": undefined,
  "generateScorePrompt": undefined,
  "preprocessPrompt": undefined,
  "preprocessResult": undefined,
  "reason": undefined,
  "run": {
    "input": [
      {
        "content": "test input",
        "role": "user",
      },
    ],
    "output": {
      "role": "assistant",
      "text": "test output",
    },
  },
  "score": 1,
}
`;

exports[`createScorer > Steps as prompt objects scorer > with preprocess and analyze prompt object 1`] = `
{
  "analyzePrompt": "Test Analyze prompt",
  "analyzeResult": {
    "inputLength": 10,
    "outputLength": 11,
  },
  "generateReasonPrompt": undefined,
  "generateScorePrompt": undefined,
  "preprocessPrompt": "Test Preprocess prompt",
  "preprocessResult": {
    "reformattedInput": "TEST INPUT",
    "reformattedOutput": "TEST OUTPUT",
  },
  "reason": undefined,
  "run": {
    "input": [
      {
        "content": "test input",
        "role": "user",
      },
    ],
    "output": {
      "role": "assistant",
      "text": "test output",
    },
  },
  "score": 1,
}
`;
